src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/JsonHubProtocol.cs

```diff
@@ -261,95 +373,139 @@ private static void WriteMessageType(JsonTextWriter writer, int type)
             writer.WriteValue(type);
         }
 
-        private InvocationMessage BindInvocationMessage(JObject json, IInvocationBinder binder)
+        private HubMessage BindCancelInvocationMessage(string invocationId)
         {
-            var invocationId = JsonUtils.GetOptionalProperty<string>(json, InvocationIdPropertyName, JTokenType.String);
-            var target = JsonUtils.GetRequiredProperty<string>(json, TargetPropertyName, JTokenType.String);
+            if (string.IsNullOrEmpty(invocationId))
+            {
+                throw new InvalidDataException($"Missing required property '{InvocationIdPropertyName}'.");
+            }
 
-            var args = JsonUtils.GetRequiredProperty<JArray>(json, ArgumentsPropertyName, JTokenType.Array);
+            return new CancelInvocationMessage(invocationId);
+        }
 
-            var paramTypes = binder.GetParameterTypes(target);
+        private HubMessage BindCompletionMessage(string invocationId, string error, object result, JToken resultToken, IInvocationBinder binder)
```

Comment:
Right now you're always passing both processed result and raw result to these bind methods which is fine but I have a different approach to float.

What do you think about processing the raw result, if it exists, when we read the invocationId? e.g. when the invocationId is read we check to see if resultToken has a value and turn the resultToken into the result then. When we come to the bind we don't need to pass both in.

I don't have a huge preference between the two.